src/bot/error.rs

```diff
@@ -0,0 +1,28 @@
+use poise::serenity_prelude as serenity;
+use thiserror::Error;
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+#[derive(Error, Debug)]
+pub enum Error {
+    /// TODO: Actually put a meaningful error in here.
+    /// needs integration with the mysql_lib module.
+    #[error("Error with the database, check logs")]
+    Database,
+    #[error("Serenity error")]
+    Serenity(#[from] serenity::Error),
+    #[error("You are not part of a semester study group")]
+    UserIsMissingSemesterStudyGroup,
+    #[error("Invalid subject id: {0}")]
+    InvalidSubjectId(i32),
+    #[error("Invalid subject name: {0}")]
+    InvalidSubjectName(String),
+}
+
+impl Error {
+    /// This can match on specific error types, so that they can be formatted using
+    /// discord-specific formatting options.
+    pub fn create_reply(&self) -> poise::CreateReply {
+        poise::CreateReply::default().content(self.to_string()).reply(true)
+    }
```

Comment:
The error reply should look similar to the old error reply (of course not 1:1). So the errors should have following fields in an embed:

- Command
- Cause
- Solution
- as a footer the command author and current time

Also, the possibility, if needed, to write something outside of the embed. (Often used for pings)

An Example how it currently looks like:

![2024-04-13_08-29](https://github.com/marcelropos/HM-DiscordBot/assets/44324946/916e2919-c29e-42b7-b46a-7eccd59f4e96)

Of course this may go over the scope of the PR so let me know how you want to handle this. I could implement it in another PR, leaving this implementation for now or I could also implement it in this PR directy. Or you may event implement it if you want.

Reply:
I think it's fine to implement it here because there's not a lot of development going on outside this, and the changes here aren't needed for other pr's right now. You can implement it in here if you want to, I'm not sure I know what would be best for users, which is 90% of the implementation.

Reply:
Ok will do it then